### PR TITLE
Refactor library (prepare to 0.9 support)

### PR DESCRIPTION
**Pull request by palkan**
Hi, guys!

I'm working on [Influxer](https://github.com/palkan/influxer) gem and going to upgrade to 0.9.
My gem depends on this one, so I've decided to upgrade `influxdb-ruby` first. 

But I found out it rather hard to work with the codebase, so I made _a little refactoring_: 
- Extract config from client
- Extract HTTP functionality to separate module
- Extract InfluxDB management functions to separate modules
- Add writer concept
- Refactor specs (add cases)
- Update README
- Add Rubocop config and _restyle_ code according to Ruby style guides

Some breaking changes:
- rename `get_<something>_list` to `list_<somethings>` (it looks less verbose);
- no more support for setting time precision on `write_point` (because it's not supported for async/udp writing, so do we really need it?);
- **drop support for Ruby < 2**.

And some new features:
- add 'denormalize' option to config / `query` to return raw InfluxDB response JSON;
- recognize `SeriesNotFound` error;
- add `Client#write_raw` method to write data without transformation (in case you want to build valid json yourself).

This is just a preparation step before upgrade to 0.9. 

@toddboom @jvshahid

### CODE REVIEW

**Comment by zlx** [comment]
:+1:  wish merge

**Comment by toddboom** [comment]
@palkan thanks - this looks awesome! i'm going to review with @kfitzpatrick tomorrow and we'll see if we can get this merged in asap.

**Comment by floored1585** [comment]
+1, any steps towards 0.9 support would be awesome!

**Comment by zanloy** [comment]
Definitely like the changes made but you should definitely break it up into smaller commits. That's a massive PR. +1 though. The sooner this gem works with 0.9, the better.

**Comment by toddboom** [comment]
@palkan i just finished cleaning up all of the v0.8.x PRs and will cut a final 0.1.9 gem. then i'll get this merged in and make 0.2.0 the first v0.9.x release.

File: lib/influxdb/client.rb [hunk 1]
```diff
@@ -41,118 +46,28 @@ class Client
...
-      when Integer
-        @retry
-      when true, nil
-        -1
-      when false
-        0
-      end
+      opts[:database] = args.first if args.first.is_a? String
+      @config = InfluxDB::Config.new(opts)
+      @stopper = false
```
Comment (yarmiganosca) [inline lib/influxdb/client.rb]: should this be `@stopped`?

File: lib/influxdb/writer/async.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    class Async
+      attr_reader :config, :client
+
+      def initialize(client, config)
+        @client = client
+        @config = config
+      end
+
+      def write(data)
+        data = data.is_a?(Array) ? data : [data]
```
Comment (yarmiganosca) [inline lib/influxdb/writer/async.rb]: `Array(data)` does exactly this. http://ruby-doc.org/core-2.0.0/Kernel.html#method-i-Array

File: lib/influxdb/client.rb [hunk 3]
```diff
@@ -41,118 +46,28 @@ class Client
...
-    def delete_shard(shard_id, server_ids)
-      data = JSON.generate({"serverIds" => server_ids})
-      delete full_url("/cluster/shards/#{shard_id}"), data
+      at_exit { stop! }
     end
 
-    def write_point(name, data, async=@async, time_precision=@time_precision)
+    # Write data point for series `name`
+    def write_point(name, data)
       data = data.is_a?(Array) ? data : [data]
```
Comment (yarmiganosca) [inline lib/influxdb/client.rb]: `Array(data)` does exactly this. If you pass it an array, it just returns that array, but if you pass it splatted arguments, it returns them as an array. http://ruby-doc.org/core-2.0.0/Kernel.html#method-i-Array

File: lib/influxdb/writer/async.rb (hunk 2 above)
Comment (palkan) [inline lib/influxdb/writer/async.rb]: I know, but: `Array(id: 1) = [:id, 1]`. So we cannot use `Array(data)`.

**Comment by palkan** [comment]
@toddboom Should I rebase this on the top of new master?

File: lib/influxdb/writer/async.rb (hunk 2 above)
Comment (yarmiganosca) [inline lib/influxdb/writer/async.rb]: Oh, I didn't even realize hashes were a possibility. My bad.

**Comment by toddboom** [comment]
@palkan if you don't mind doing the rebase, it will save me some time - i'm planning to get this merged in later this afternoon, so let me know if you're working on it! thanks!

**Comment by palkan** [comment]
@toddboom I working on it.

**Comment by toddboom** [comment]
@palkan it looks like only `jruby-head` failed, and that one looks like a fluke - i'll get this merged in tonight! thanks for the quick turnaround!

**Comment by palkan** [comment]
@toddboom Oh, I haven't finished _rebasing_ yet. I still have to check a dozen of commits to ensure that all newly added functionality works (e.g. some specs are missing). I'll make another PR.

**Pull request merged**